packages/ampd-sdk/src/grpc/connection_pool.rs

```diff
@@ -0,0 +1,394 @@
+/// - The pool handles connection retries and failure recovery
+#[derive(Debug)]
+pub struct ConnectionPool {
+    url: Url,
+    connection_state: watch::Sender<ConnectionState>,
+    message_receiver: mpsc::Receiver<ClientMessage>,
+    retry_policy: RetryPolicy,
+    is_reconnecting: bool,
+}
+
+impl ConnectionPool {
+    pub fn new(url: Url) -> (Self, ConnectionHandle) {
+        let (state_sender, state_receiver) = watch::channel(ConnectionState::Disconnected);
+        let (msg_sender, msg_receiver) = mpsc::channel(MESSAGE_RECEIVER_CHANNEL_CAPACITY);
+
+        let handle = ConnectionHandle {
+            connection_receiver: state_receiver,
+            message_sender: msg_sender,
+        };
+
+        let pool = Self {
+            url,
+            connection_state: state_sender,
+            message_receiver: msg_receiver,
+            retry_policy: DEFAULT_RETRY_POLICY,
+            is_reconnecting: false,
+        };
+
+        (pool, handle)
+    }
+
+    pub async fn run(mut self, token: CancellationToken) -> Result<(), Error> {
+        if let Err(e) = self.connect().await {
+            warn!(err = ?e, "initial connection failed, will retry on client requests");
+        }
+
+        loop {
+            tokio::select! {
+                message = self.message_receiver.recv() => {
+                    match message {
+                        Some(ClientMessage::ConnectionFailed(details)) => {
+                            warn!("client reported connection failure: {}", details);
+                            if !self.is_reconnecting {
+                                if let Err(e) = self.handle_connection_failure().await {
+                                    warn!(err = ?e, "reconnection attempt failed, will retry on next client request");
+                                }
+                            }
+                        }
+                        None => break,
+                    }
+                }
+                _ = token.cancelled() => {
+                    info!("connection pool shutting down due to cancellation");
+                    break;
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn connect(&self) -> Result<(), Error> {
+        let endpoint: transport::Endpoint = self.url.as_str().parse().into_report()?;
+        let endpoint = endpoint
+            .connect_timeout(DEFAULT_INITIAL_TIMEOUT)
+            .timeout(DEFAULT_RPC_TIMEOUT)
+            .keep_alive_timeout(KEEPALIVE_TIMEOUT)
+            .keep_alive_while_idle(KEEPALIVE_WHILE_IDLE)
+            .http2_keep_alive_interval(KEEPALIVE_TIME);
+
+        match endpoint.connect().await {
+            Ok(channel) => {
+                self.notify_clients(ConnectionState::Connected(channel));
+                Ok(())
+            }
+            Err(status) => Err(status).into_report(),
+        }
+    }
+
+    async fn handle_connection_failure(&mut self) -> Result<(), Error> {
```

Comment:
simplify your naming
```suggestion
    async fn reconnect(&mut self) -> Result<(), Error> {
```